ngraph/core/reference/src/runtime/reference/einsum.cpp

```diff
@@ -0,0 +1,1239 @@
+                        new HostTensor(input_ptr->get_element_type(), identity_shape));
+
+                    T* identity_data_ptr = identity->get_data_ptr<T>();
+                    size_t data_size =
+                        shape_size(identity_shape) * identity->get_element_type().size();
+                    std::memset(identity_data_ptr, 0, data_size);
+
+                    // Identity[k,k,...,k] element is placed in k*p^(n-1) + ... + k*p + k position,
+                    // where p is a size of one Identity dimension,
+                    // n is occurence number for the considered label and k in [0; p).
+                    // Note that k*p^(n-1) + ... + k*p + k = k * (p^n-1)/(p-1) = k * alpha
+                    size_t p = repeated_label_dim_size;
+                    if (p == 1)
+                    {
+                        identity_data_ptr[0] = static_cast<T>(1);
+                        return identity;
+                    }
+
+                    size_t n = repeated_label_dims.size();
+                    size_t alpha = (static_cast<size_t>(std::pow(p, n)) - 1) / (p - 1);
+                    size_t offset = 0;
+                    for (size_t k = 0; k < p; ++k)
+                    {
+                        identity_data_ptr[offset] = static_cast<T>(1);
+                        offset += alpha;
+                    }
+
+                    return identity;
+                }
+
+                /// \brief      Computes a multiplication of Identity tensors built for each
+                /// repeated label
+                ///
+                template <typename T>
+                HostTensorPtr build_multi_identity(
+                    const HostTensorPtr& input_ptr,
+                    const std::vector<std::string>& repeated_labels,
+                    std::unordered_map<std::string, std::vector<size_t>>& label_dim_map)
+                {
+                    Shape input_shape = input_ptr->get_shape();
+
+                    // initially set multi-identity with identity for the first repeated label
+                    NGRAPH_CHECK(repeated_labels.size() > 0);
+                    auto first_repeated_label = repeated_labels[0];
+                    NGRAPH_CHECK(label_dim_map.find(first_repeated_label) != label_dim_map.end());
+                    auto repeated_label_dims = label_dim_map[first_repeated_label];
+                    HostTensorPtr multi_identity =
+                        build_identity<T>(input_ptr, repeated_label_dims);
+
+                    for (size_t label_ind = 1; label_ind < repeated_labels.size(); ++label_ind)
+                    {
+                        NGRAPH_CHECK(label_dim_map.find(repeated_labels[label_ind]) !=
+                                     label_dim_map.end());
+                        repeated_label_dims = label_dim_map[repeated_labels[label_ind]];
+                        HostTensorPtr identity = build_identity<T>(input_ptr, repeated_label_dims);
```

Comment:
Looks like this `HostTensor` should be equal to `multi_identity` from 645 line. Can we reuse tensors which contain same data?

Reply:
I found only one place where it is possible to re-use because shapes of tensors for intermediate results are no changes between iterations.